File: my-folder/0567-permutation-in-string/solution.py
class Solution(object):
    def checkInclusion(self, s1, s2):
        """
        :type s1: str
        :type s2: str
        :rtype: bool
        """
        c1={}
        for i in s1:
            c1[i]=1+c1.get(i,0)
        k=len(s1)
        c2={}
        L=0
        for R in range(len(s2)):
            c2[s2[R]]=1+c2.get(s2[R],0)
            while(R-L+1>k):
                c2[s2[L]]-=1
                if(c2[s2[L]]==0):
                    del c2[s2[L]]
                L+=1
            if(c1==c2):
                return True
        return False


        
        

File: my-folder/0567-permutation-in-string/solution.java
class Solution {
    public boolean checkInclusion(String s1, String s2) {
        if (s1.length() > s2.length()) return false;

        HashMap<Character, Integer> s1Hash = new HashMap<>();
        HashMap<Character, Integer> window = new HashMap<>();
    
        for(char c: s1.toCharArray()){
            s1Hash.put(c, s1Hash.getOrDefault(c, 0) + 1);
        }

        int left = 0;
        for(int right = 0; right < s2.length(); right++){
            char curr = s2.charAt(right);
            window.put(curr, window.getOrDefault(curr, 0) + 1);
            while(right - left + 1 > s1.length()){
                char leftChar = s2.charAt(left);
                window.put(leftChar, window.get(leftChar) - 1);
                if (window.get(leftChar) == 0) {
                    window.remove(leftChar);
                }
                left++;
            }

            if(window.equals(s1Hash)) return true;
            
        }
        
        return false;
    }
}
